R/run.R

```diff
@@ -304,51 +310,50 @@ CmdStanRun$set("private", name = "run_variational_", value = .run_other)
 
 # CmdStanProcs ------------------------------------------------------------
 
-# System processes for model fitting
-# Currently only used to run sampling in parallel
+# System processes for model fitting 
 CmdStanProcs <- R6::R6Class(
   classname = "CmdStanProcs",
   public = list(
-    # @param num_runs The number of CmdStan runs. For MCMC this is the number of
-    #   chains. Currently for other methods this must be set to 1.
-    # @param num_cores The number of cores for running MCMC chains in parallel.
-    #   Currently for other method this must be set to 1.
-    # @param threads_per_chain The number of threads to use per MCMC chain
+    # @param num_procs The number of CmdStan processes to start for a run. 
+    #   For MCMC this is the number of chains. Currently for other methods
+    #   this must be set to 1.
+    # @param parallel_procs The maximum number of processes to run in parallel.
+    #   Currently for non-sampling this must be set to 1.
+    # @param threads_per_proc The number of threads to use per MCMC chain
     #   to run parallel sections of model.
-    initialize = function(num_runs, num_cores, threads_per_chain = NULL) {
-      checkmate::assert_integerish(num_runs, lower = 1, len = 1, any.missing = FALSE)
-      checkmate::assert_integerish(num_cores, lower = 1, len = 1, any.missing = FALSE,
-                                   .var.name = "parallel_chains")
-      checkmate::assert_integerish(threads_per_chain, lower = 1, len = 1, null.ok = TRUE,
-                                   .var.name = "threads_per_chain")
-      private$num_runs_ <- as.integer(num_runs)
-      private$num_cores_ <- as.integer(num_cores)
-      private$threads_per_chain_ <- threads_per_chain
-      private$active_cores_ <- 0
-      private$run_ids_ <- seq_len(num_runs)
-      zeros <- rep(0, num_runs)
-      private$chain_info_ <- data.frame(
-        id = private$run_ids_,
-        state = zeros,
-        start_time = zeros,
-        warmup_time = zeros,
-        sampling_time = zeros,
-        total_time = zeros,
-        last_section_start_time = zeros
-      )
+    initialize = function(num_procs, parallel_procs = NULL, threads_per_proc = NULL) {
+      checkmate::assert_integerish(num_procs, lower = 1, len = 1, any.missing = FALSE)
+      checkmate::assert_integerish(parallel_procs, lower = 1, len = 1, any.missing = FALSE,
+                                   .var.name = "parallel_procs", null.ok = TRUE)
+      checkmate::assert_integerish(threads_per_proc, lower = 1, len = 1, null.ok = TRUE,
+                                   .var.name = "threads_per_proc")
+      private$num_procs_ <- as.integer(num_procs)
+      if (is.null(parallel_procs)) {
+        private$parallel_procs_ <- private$num_procs_
+      } else {
+        private$parallel_procs_ <- as.integer(parallel_procs)
+      }
+      private$threads_per_proc_ <- threads_per_proc
```

Comment:
do we need `as.integer(threads_per_proc)`? or just for `parallel_procs`?

Reply:
Yep. Good call.